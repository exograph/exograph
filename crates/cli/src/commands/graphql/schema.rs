// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use anyhow::Result;
use async_trait::async_trait;
use clap::{builder::PossibleValue, Arg, Command, ValueEnum};

use std::{
    fs::{self, File},
    io::Write,
    path::{Path, PathBuf},
};

use crate::commands::{
    command::{default_model_file, get, output_arg, CommandDefinition},
    schema::util::create_system,
    util::use_ir_arg,
};
use crate::config::Config;

pub(super) struct SchemaCommandDefinition {}

#[async_trait]
impl CommandDefinition for SchemaCommandDefinition {
    fn command(&self) -> clap::Command {
        Command::new("schema")
            .about("Obtain GraphQL schema")
            .arg(output_arg().long_help(
                "Output file for the introspection result. Default: generated/schema.json",
            ))
            .arg(
                Arg::new("format")
                    .long("format")
                    .short('f')
                    .value_parser(clap::builder::EnumValueParser::<SchemaFormat>::new())
                    .help("Output format. Default: graphql (sdl)")
                    .default_value("graphql"),
            )
            .arg(use_ir_arg())
    }

    /// Create a database schema from a exograph model
    async fn execute(&self, matches: &clap::ArgMatches, _config: &Config) -> Result<()> {
        let use_ir: bool = matches.get_flag("use-ir");

        let model_path: PathBuf = default_model_file();

        let serialized_system = create_system(&model_path, None, use_ir).await?;

        let introspection_result = testing::get_introspection_result(serialized_system).await?;

        let format: SchemaFormat = match get(matches, "format") {
            Some(format) => format,
            None => SchemaFormat::Graphql,
        };

        match format {
            SchemaFormat::Json => {
                let output: PathBuf = match get(matches, "output") {
                    Some(output) => output,
                    None => {
                        fs::create_dir_all("generated")?;
                        let output = Path::new("generated/schema.json").to_path_buf();
                        output
                    }
                };

                serde_json::to_writer_pretty(&mut File::create(output)?, &introspection_result)?;
            }
            SchemaFormat::Graphql => {
                let output: PathBuf = match get(matches, "output") {
                    Some(output) => output,
                    None => {
                        fs::create_dir_all("generated")?;
                        let output = Path::new("generated/schema.graphql").to_path_buf();
                        output
                    }
                };

                let schema_string = testing::schema_sdl(introspection_result).await?;

                File::create(output)?.write_all(schema_string.as_bytes())?;
            }
        }

        Ok(())
    }
}

#[derive(Clone, Debug)]
enum SchemaFormat {
    Json,
    Graphql,
}

impl ValueEnum for SchemaFormat {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::Json, Self::Graphql]
    }

    fn to_possible_value(&self) -> Option<PossibleValue> {
        match self {
            Self::Json => Some(PossibleValue::new("json")),
            Self::Graphql => Some(PossibleValue::new("graphql")),
        }
    }
}
