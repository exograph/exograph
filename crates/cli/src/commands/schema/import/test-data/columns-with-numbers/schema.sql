CREATE TABLE "metrics" (
        "id" SERIAL PRIMARY KEY,
        "name" TEXT NOT NULL,
        "min_30d_price" REAL NOT NULL,
        "max30d_price" REAL NOT NULL
);