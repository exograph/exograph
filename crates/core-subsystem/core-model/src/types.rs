// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Common support for representing GraphQL types such as `Int`, `List<Int>`, `Optional<Int>`, `Optional<List<Int>>`, etc.
//!

use async_graphql_parser::types::{BaseType, Type};
use async_graphql_value::Name;
use serde::{Deserialize, Serialize};

use crate::mapped_arena::{SerializableSlab, SerializableSlabIndex};

/// A type that can be used as a type for fields and return types
/// Currently supports only list and optional decorations
#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum FieldType<T> {
    Plain(T),
    List(Box<FieldType<T>>),
    Optional(Box<FieldType<T>>),
}

pub trait Named {
    fn name(&self) -> &str;
}

impl<T> FieldType<T> {
    /// Return the inner type, if any
    pub fn inner(&self) -> Option<&FieldType<T>> {
        match self {
            FieldType::Plain(_) => None,
            FieldType::List(inner) | FieldType::Optional(inner) => Some(inner),
        }
    }

    /// Return the base type (i.e. by removing optional decoration)
    pub fn base_type(&self) -> &FieldType<T> {
        match self {
            FieldType::Optional(inner_typ) => inner_typ.as_ref(),
            _ => self,
        }
    }

    /// Return the innermost (i.e. leaf) type
    pub fn innermost(&self) -> &T {
        match self {
            FieldType::Plain(inner) => inner,
            FieldType::List(inner) | FieldType::Optional(inner) => inner.innermost(),
        }
    }

    /// Wrap the destination type to match the structure of the source type
    pub fn wrap<D>(&self, dest: D) -> FieldType<D> {
        match self {
            FieldType::Plain(_) => FieldType::Plain(dest),
            FieldType::Optional(optional) => FieldType::Optional(Box::new(optional.wrap(dest))),
            FieldType::List(list) => FieldType::List(Box::new(list.wrap(dest))),
        }
    }
}

impl<T: Clone> FieldType<T> {
    /// Compute the optional version of the given type
    pub fn optional(&self) -> Self {
        match self {
            FieldType::Optional(_) => self.clone(), // Already optional
            _ => FieldType::Optional(Box::new(self.clone())),
        }
    }
}

/// Transforms the type into an introspection type
///
/// The complexity of this function is due to the fact that the GraphQL spec and hence the
/// introspection type ([`Type`]) does not support nested optionals. However, [`FieldType`]
/// being more general, does. This function will panic if it encounters a nested optional.
impl<T: Named> From<&FieldType<T>> for Type {
    fn from(ft: &FieldType<T>) -> Self {
        match ft {
            FieldType::Plain(base) => Type {
                base: BaseType::Named(Name::new(base.name())),
                nullable: false,
            },
            FieldType::Optional(underlying) => {
                let Type { base, nullable } = underlying.as_ref().into();

                if nullable {
                    panic!("Optional type cannot be nested")
                }
                Type {
                    base,
                    nullable: true,
                }
            }
            FieldType::List(underlying) => Type {
                base: BaseType::List(Box::new(underlying.as_ref().into())),
                nullable: false,
            },
        }
    }
}

impl<T: Named> Named for FieldType<T> {
    fn name(&self) -> &str {
        match self {
            FieldType::Plain(plain) => plain.name(),
            FieldType::List(list) => list.name(),
            FieldType::Optional(optional) => optional.name(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct BaseOperationReturnType<T> {
    pub associated_type_id: SerializableSlabIndex<T>,
    pub type_name: String,
}

impl<T> Named for BaseOperationReturnType<T> {
    fn name(&self) -> &str {
        &self.type_name
    }
}

pub type OperationReturnType<T> = FieldType<BaseOperationReturnType<T>>;

impl<T> OperationReturnType<T> {
    pub fn typ<'a>(&'a self, types: &'a SerializableSlab<T>) -> &T {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType {
                associated_type_id, ..
            }) => &types[*associated_type_id],
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.typ(types)
            }
        }
    }

    pub fn type_name(&self) -> &str {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType { type_name, .. }) => type_name,
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.type_name()
            }
        }
    }
}
