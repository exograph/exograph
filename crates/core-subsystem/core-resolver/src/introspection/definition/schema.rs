// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::collections::{HashMap, HashSet};

use async_graphql_parser::{
    types::{
        BaseType, FieldDefinition, InputValueDefinition, ObjectType, Type, TypeDefinition, TypeKind,
    },
    Positioned,
};

use async_graphql_value::Name;
use core_model::type_normalization::{
    default_positioned, default_positioned_name, TypeDefinitionIntrospection,
};

use crate::{plugin::SubsystemResolver, validation::underlying_type};

#[derive(Debug, Clone)]
pub struct Schema {
    pub type_definitions: Vec<TypeDefinition>,
    pub(crate) schema_field_definition: FieldDefinition,
    pub(crate) type_field_definition: FieldDefinition,
}

pub const QUERY_ROOT_TYPENAME: &str = "Query";
pub const MUTATION_ROOT_TYPENAME: &str = "Mutation";
pub const SUBSCRIPTION_ROOT_TYPENAME: &str = "Subscription";

impl Schema {
    pub fn new_from_resolvers(
        subsystem_resolvers: &[Box<dyn SubsystemResolver + Send + Sync>],
    ) -> Schema {
        let type_definitions: Vec<TypeDefinition> = {
            let mut typedefs = subsystem_resolvers
                .iter()
                .fold(vec![], |mut acc, resolver| {
                    acc.extend(resolver.schema_types());
                    acc
                });

            // ensure introspection outputs fields in a stable order
            typedefs.sort_by_key(|f| f.name.clone());
            typedefs
        };

        let queries = {
            let mut queries = subsystem_resolvers
                .iter()
                .fold(vec![], |mut acc, resolver| {
                    acc.extend(resolver.schema_queries());
                    acc
                });

            // ensure introspection outputs queries in a stable order
            queries.sort_by_key(|q| q.name.clone());
            queries
        };

        let mutations = {
            let mut mutations = subsystem_resolvers
                .iter()
                .fold(vec![], |mut acc, resolver| {
                    acc.extend(resolver.schema_mutations());
                    acc
                });

            // ensure introspection outputs mutations in a stable order
            mutations.sort_by_key(|m| m.name.clone());
            mutations
        };

        Self::new(type_definitions, queries, mutations)
    }

    pub fn new(
        type_definitions: Vec<TypeDefinition>,
        queries: Vec<FieldDefinition>,
        mutations: Vec<FieldDefinition>,
    ) -> Schema {
        let mut type_definitions = type_definitions;

        // Ideally, we should surround it with `if !queries.is_empty() {` (like we do for mutations next)
        // but GraphQL spec requires a `Query` type to be present in the schema.
        // https://spec.graphql.org/June2018/#sec-Root-Operation-Types
        // "The query root operation type must be provided and must be an Object type."
        // So we always add a `Query` type to the schema. This means introspection will fail
        // if there are no queries. See https://github.com/exograph/exograph/issues/480
        //
        // Even though we resolve __type and __schema fields for the Query
        // type, GraphQL spec doesn't allow them to be exposed as an
        // ordinary field. Therefore, we have to treat them specially (see
        // SelectionSetValidator::validate_field)
        type_definitions.push(TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name(QUERY_ROOT_TYPENAME),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields: queries.into_iter().map(default_positioned).collect(),
            }),
        });

        if !mutations.is_empty() {
            type_definitions.push(TypeDefinition {
                extend: false,
                description: None,
                name: default_positioned_name(MUTATION_ROOT_TYPENAME),
                directives: vec![],
                kind: TypeKind::Object(ObjectType {
                    implements: vec![],
                    fields: mutations.into_iter().map(default_positioned).collect(),
                }),
            });
        };

        type_definitions.push(Self::create_schema_type_definition());
        type_definitions.push(Self::create_type_definition());
        type_definitions.push(Self::create_field_definition());
        type_definitions.push(Self::create_directive_definition());
        type_definitions.push(Self::create_directive_location_definition());
        type_definitions.push(Self::create_input_value_definition());

        // We may have unused scalars (an artifact of how our typechecking starts with supplying all scalars)
        // So here we remove unused ones
        let (scalars, mut type_definitions): (Vec<_>, Vec<_>) = type_definitions
            .into_iter()
            .partition(|td| matches!(td.kind, TypeKind::Scalar));

        // Start with assuming no scalars are used
        let mut unused_scalars_names: HashSet<String> = scalars
            .iter()
            .map(|td| td.name.node.as_str().to_string())
            .collect();

        // Next, remove scalars that are used
        for td in &type_definitions {
            match &td.kind {
                TypeKind::Object(object_type) => {
                    for field in &object_type.fields {
                        unused_scalars_names.remove(underlying_type(&field.node.ty.node).as_str());
                    }
                }
                TypeKind::Interface(interface_type) => {
                    for field in &interface_type.fields {
                        unused_scalars_names.remove(underlying_type(&field.node.ty.node).as_str());
                    }
                }
                TypeKind::InputObject(input_object_type) => {
                    for field in &input_object_type.fields {
                        unused_scalars_names.remove(underlying_type(&field.node.ty.node).as_str());
                    }
                }
                _ => {}
            }
        }

        let used_scalars = scalars
            .into_iter()
            .filter(|td| !unused_scalars_names.contains(td.name.node.as_str()));
        // Create a unique list of scalars (each subsystem may expose the same scalar)
        let unique_scalars: HashMap<_, _> = used_scalars
            .into_iter()
            .map(|td| (td.name.node.as_str().to_string(), td))
            .collect();

        type_definitions.extend(unique_scalars.into_values());

        Schema {
            type_definitions,
            schema_field_definition: Self::create_field(
                "__schema",
                false,
                Some("Access the current type schema of this server."),
                "__Schema",
                vec![],
            )
            .node,
            type_field_definition: Self::create_field(
                "__type",
                true,
                None,
                "__Type",
                vec![default_positioned(InputValueDefinition {
                    description: None,
                    name: default_positioned_name("name"),
                    directives: vec![],
                    default_value: None,
                    ty: default_positioned(Type {
                        base: BaseType::Named(Name::new("String")),
                        nullable: true,
                    }),
                })],
            )
            .node,
        }
    }

    pub fn get_type_definition(&self, type_name: &str) -> Option<&TypeDefinition> {
        self.type_definitions
            .iter()
            .find(|td| td.name().as_str() == type_name)
    }

    fn create_schema_type_definition() -> TypeDefinition {
        let directives_field = Self::create_list_field(
            "directives",
            false,
            Some("A list of the directives supported by this server."),
            "__Directive",
            vec![],
        );

        let types_field = Self::create_list_field(
            "types",
            false,
            Some("A list of the types supported by this server."),
            "__Type",
            vec![],
        );

        let mut fields: Vec<_> = ["queryType", "mutationType", "subscriptionType"]
            .into_iter()
            .map(|field_name| {
                default_positioned(FieldDefinition {
                    description: None,
                    name: default_positioned_name(field_name),
                    arguments: vec![],
                    ty: default_positioned(Type {
                        base: BaseType::Named(Name::new("__Type")),
                        nullable: false,
                    }),
                    directives: vec![],
                })
            })
            .collect();

        fields.push(Self::create_field(
            "description",
            true,
            None,
            "String",
            vec![],
        ));
        fields.push(directives_field);
        fields.push(types_field);

        TypeDefinition {
            extend: false,
            description: Some(default_positioned(
                "The current type schema of this server.".to_string(),
            )),
            name: default_positioned_name("__Schema"),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields,
            }),
        }
    }

    fn create_type_definition() -> TypeDefinition {
        let fields_arguments = vec![default_positioned(InputValueDefinition {
            description: None,
            name: default_positioned_name("includeDeprecated"),
            directives: vec![],
            default_value: None,
            ty: default_positioned(Type {
                base: BaseType::Named(Name::new("Boolean")),
                nullable: true,
            }),
        })];

        let fields = vec![
            Self::create_field("name", true, None, "String", vec![]),
            Self::create_field("description", true, None, "String", vec![]),
            Self::create_field("kind", false, None, "String", vec![]),
            Self::create_field("specifiedByURL", true, None, "String", vec![]),
            Self::create_field("ofType", false, None, "__Type", vec![]),
            Self::create_list_field("fields", false, None, "__Field", fields_arguments.clone()),
            Self::create_list_field(
                "inputFields",
                false,
                None,
                "__InputValue",
                fields_arguments.clone(),
            ),
            Self::create_list_field("enumValues", false, None, "__InputValue", fields_arguments),
            Self::create_list_field("interfaces", false, None, "__Type", vec![]),
            Self::create_list_field("possibleTypes", false, None, "__Type", vec![]),
        ];

        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name("__Type"),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields,
            }),
        }
    }

    fn create_field_definition() -> TypeDefinition {
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name("__Field"),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields: vec![
                    Self::create_field("name", true, None, "String", vec![]),
                    Self::create_field("description", true, None, "String", vec![]),
                    Self::create_list_field("args", false, None, "__InputValue", vec![]),
                    Self::create_field("type", false, None, "__Type", vec![]),
                    Self::create_field("isDeprecated", false, None, "Boolean", vec![]),
                    Self::create_field("deprecationReason", true, None, "String", vec![]),
                ],
            }),
        }
    }

    fn create_directive_definition() -> TypeDefinition {
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name("__Directive"),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields: vec![
                    Self::create_field("name", false, None, "String", vec![]),
                    Self::create_field("description", true, None, "String", vec![]),
                    Self::create_field("isRepeatable", false, None, "Boolean", vec![]),
                    Self::create_list_field("args", false, None, "__InputValue", vec![]),
                    Self::create_list_field(
                        "locations",
                        false,
                        None,
                        "__DirectiveLocation",
                        vec![],
                    ),
                ],
            }),
        }
    }

    fn create_directive_location_definition() -> TypeDefinition {
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name("__DirectiveLocation"),
            directives: vec![],
            kind: TypeKind::Scalar,
        }
    }

    fn create_input_value_definition() -> TypeDefinition {
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name("__InputValue"),
            directives: vec![],
            kind: TypeKind::Object(ObjectType {
                implements: vec![],
                fields: vec![
                    Self::create_field("name", false, None, "String", vec![]),
                    Self::create_field("description", true, None, "String", vec![]),
                    Self::create_field("type", false, None, "__Type", vec![]),
                    Self::create_field("defaultValue", true, None, "String", vec![]),
                    Self::create_field("isDeprecated", false, None, "Boolean", vec![]),
                    Self::create_field("deprecationReason", true, None, "String", vec![]),
                ],
            }),
        }
    }

    pub fn create_field(
        name: &str,
        nullable: bool,
        description: Option<&str>,
        element_type: &str,
        arguments: Vec<Positioned<InputValueDefinition>>,
    ) -> Positioned<FieldDefinition> {
        default_positioned(FieldDefinition {
            description: description.map(|d| default_positioned(d.to_string())),
            name: default_positioned_name(name),
            arguments,
            ty: default_positioned(Type {
                base: BaseType::Named(Name::new(element_type)),
                nullable,
            }),
            directives: vec![],
        })
    }

    fn create_list_field(
        name: &str,
        nullable: bool,
        description: Option<&str>,
        element_type: &str,
        arguments: Vec<Positioned<InputValueDefinition>>,
    ) -> Positioned<FieldDefinition> {
        default_positioned(FieldDefinition {
            description: description.map(|d| default_positioned(d.to_string())),
            name: default_positioned_name(name),
            arguments,
            ty: default_positioned(Type {
                base: BaseType::List(Box::new(Type {
                    base: BaseType::Named(Name::new(element_type)),
                    nullable,
                })),
                nullable,
            }),
            directives: vec![],
        })
    }
}
