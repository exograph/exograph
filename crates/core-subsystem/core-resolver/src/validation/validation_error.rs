use async_graphql_parser::Pos;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ValidationError {
    #[error("{0}")]
    QueryParsingFailed(String, Pos, Option<Pos>),

    #[error("Variable '{0}' not found")]
    VariableNotFound(String, Pos),

    #[error("Variable '{0}' could not be deserialized: {2}")]
    MalformedVariable(String, Pos, serde_json::Error),

    #[error("Fragment definition '{0}' not found")]
    FragmentDefinitionNotFound(String, Pos),

    #[error("Inline fragments are not supported")]
    InlineFragmentNotSupported(Pos),

    #[error("No such operation '{0}'")]
    OperationNotFound(String, Pos),

    #[error("Field '{0}' is not valid for type '{1}'")]
    InvalidField(String, String, Pos),

    #[error("Field '{0}' is of a scalar type, which should not specify fields")]
    ScalarWithField(String, Pos),

    #[error("Field type '{0}' is not valid")]
    InvalidFieldType(String, Pos),

    #[error("Required argument '{0}' not found")]
    RequiredArgumentNotFound(String, Pos),

    #[error("Argument(s) '{0:?}' invalid for '{1}'")]
    StrayArguments(Vec<String>, String, Pos),

    #[error(
        "Argument '{argument_name}' is not of a valid type. Expected '{expected_type}', got '{actual_type}'"
    )]
    InvalidArgumentType {
        argument_name: String,
        expected_type: String,
        actual_type: String,
        pos: Pos,
    },

    #[error("No operation found")]
    NoOperationFound,

    #[error("Must provide operation name if query contains multiple operations")]
    MultipleOperationsNoOperationName,

    #[error("operationName '{0}' doesn't match any operation")]
    MultipleOperationsUnmatchedOperationName(String),
}

impl ValidationError {
    pub fn position1(&self) -> Pos {
        match self {
            ValidationError::QueryParsingFailed(_, pos, _) => *pos,
            ValidationError::VariableNotFound(_, pos) => *pos,
            ValidationError::MalformedVariable(_, pos, _) => *pos,
            ValidationError::FragmentDefinitionNotFound(_, pos) => *pos,
            ValidationError::InlineFragmentNotSupported(pos) => *pos,
            ValidationError::OperationNotFound(_, pos) => *pos,
            ValidationError::InvalidField(_, _, pos) => *pos,
            ValidationError::InvalidFieldType(_, pos) => *pos,
            ValidationError::ScalarWithField(_, pos) => *pos,
            ValidationError::RequiredArgumentNotFound(_, pos) => *pos,
            ValidationError::StrayArguments(_, _, pos) => *pos,
            ValidationError::NoOperationFound => Pos::default(),
            ValidationError::MultipleOperationsNoOperationName => Pos::default(),
            ValidationError::MultipleOperationsUnmatchedOperationName(_) => Pos::default(),
            ValidationError::InvalidArgumentType { pos, .. } => *pos,
        }
    }

    pub fn position2(&self) -> Option<Pos> {
        match self {
            ValidationError::QueryParsingFailed(_, _, pos) => *pos,
            _ => None,
        }
    }
}
