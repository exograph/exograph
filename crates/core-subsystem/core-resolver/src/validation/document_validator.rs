use crate::validation::validation_error::ValidationError;
use async_graphql_parser::types::{DocumentOperations, ExecutableDocument};
use async_graphql_value::Name;
use serde_json::{Map, Value};
use tracing::instrument;

use crate::introspection::definition::schema::Schema;

use super::{operation::ValidatedOperation, operation_validator::OperationValidator};

/// Context for validating a document.
pub struct DocumentValidator<'a> {
    schema: &'a Schema,
    operation_name: Option<String>,
    variables: Option<Map<String, Value>>,
}

impl<'a> DocumentValidator<'a> {
    pub fn new(
        schema: &'a Schema,
        operation_name: Option<String>,
        variables: Option<Map<String, Value>>,
    ) -> Self {
        Self {
            schema,
            operation_name,
            variables,
        }
    }

    /// Validate the query payload.
    ///
    /// Validations performed:
    /// - Validate that either there is only one operation or the operation name specified matches one of the operations in the document
    /// - Validate that there is at least one operation
    /// - Other validations are delegated to the operation validator
    #[instrument(
        name = "DocumentValidator::validate"
        skip(self, document)
        )]
    pub fn validate(
        self,
        document: ExecutableDocument,
    ) -> Result<ValidatedOperation, ValidationError> {
        let (operation_name, raw_operation) = match document.operations {
            DocumentOperations::Single(operation) => Ok((self.operation_name, operation)),
            DocumentOperations::Multiple(mut operations) => {
                if operations.is_empty() {
                    Err(ValidationError::NoOperationFound)
                } else {
                    match self.operation_name {
                        None if operations.len() == 1 => {
                            // Per https://graphql.org/learn/queries/#operation-name, `operationName` is required
                            // only for multiple operations, but async-graphql parses a named operation (`query Foo { ... }`)
                            // to `DocumentOperations::Multiple` even if there is only one operation. So we add an additional
                            // check here to make sure that the operation name is enforced only for truly multiple operations.

                            // This unwrap is okay because we already check that there is exactly one operation.
                            let (operation_name, operation) =
                                operations.into_iter().next().unwrap();
                            Ok((Some(operation_name.to_string()), operation))
                        }
                        None => Err(ValidationError::MultipleOperationsNoOperationName),
                        Some(operation_name) => {
                            let operation = operations.remove(&Name::new(&operation_name));

                            match operation {
                                None => {
                                    Err(ValidationError::MultipleOperationsUnmatchedOperationName(
                                        operation_name,
                                    ))
                                }
                                Some(operation) => Ok((Some(operation_name), operation)),
                            }
                        }
                    }
                }
            }
        }?;

        let operation_validator = OperationValidator::new(
            self.schema,
            operation_name,
            self.variables,
            document.fragments,
        );

        operation_validator.validate(raw_operation)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_graphql_parser::parse_query;
    use core_plugin_shared::{
        serializable_system::SerializableSystem, system_serializer::SystemSerializer,
    };

    #[test]
    fn argument_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1) {
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn with_operation_name_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query ConcertById {
                concert(id: 1) {
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn stray_argument_invalid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1, foo: "bar") { # foo is not a valid argument
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn unspecified_required_argument_invalid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert { # id argument is required here
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn variable_resolution_valid() {
        let schema = create_test_schema();

        let variables = create_variables(
            r#"
            {
                "concert_id": 1,
                "venue_id": 2
            }"#,
        );

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: Some(variables),
        };

        let query = r#"
            query($concert_id: Int!, $venue_id: Int!) {
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn variable_resolution_invalid() {
        let schema = create_test_schema();

        let variables = create_variables(r#"{ "concert_id": 2 }"#);
        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: Some(variables),
        };

        let query = r#"
            query($concert_id: Int!, $venue_id: Int!) { # venue_id is not a specified in variables
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn invalid_subfield() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1) {
                    id
                    title
                    foobar
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn aliases_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                firstConcert: concert(id: 1) {
                    id
                    headLine: title
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn multi_operations_valid() {
        let schema = create_test_schema();

        let query = r#"
            query concert1 {
                concert(id: 1) {
                    id
                    headLine: title
                }
            }

            query concert2 {
                concert(id: 2) {
                    id
                    headLine: title
                }
            }
        "#;

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: Some("concert1".to_string()),
            variables: None,
        };

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: Some("concert2".to_string()),
            variables: None,
        };

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn multi_operations_no_operation_name_invalid() {
        let schema = create_test_schema();

        let query = r#"
            query concert1 {
                concert(id: 1) {
                    id
                    headLine: title
                }
            }

            query concert2 {
                concert(id: 2) {
                    id
                    headLine: title
                }
            }
        "#;

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn multi_operations_mismatched_operation_name_invalid() {
        let schema = create_test_schema();

        let query = r#"
            query concert1 {
                concert(id: 1) {
                    id
                    headLine: title
                }
            }

            query concert2 {
                concert(id: 2) {
                    id
                    headLine: title
                }
            }
        "#;

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: Some("foo".to_string()),
            variables: None,
        };

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    fn create_variables(variables: &str) -> Map<String, Value> {
        serde_json::from_str(variables).unwrap()
    }

    fn create_test_schema() -> Schema {
        let test_clay = r#"
            model Concert {
                id: Int = autoincrement() @pk
                title: String
                venue: Venue
            }

            model Venue {
                id: Int = autoincrement() @pk
                name: String
                concerts: Set<Concert>
            }
        "#;
        let postgres_subsystem =
            create_postgres_system_from_str(test_clay, "test.clay".to_string());

        Schema::new(
            postgres_subsystem.schema_types(),
            postgres_subsystem.schema_queries(),
            postgres_subsystem.schema_mutations(),
        )
    }

    fn create_query_document(query_str: &str) -> ExecutableDocument {
        parse_query(query_str).unwrap()
    }

    // TODO: Rethink this approach, where core (even though it's a test) depends on a specific subsystem (database)

    use postgres_model::model::ModelPostgresSystem;
    pub fn create_postgres_system_from_str(
        model_str: &str,
        file_name: String,
    ) -> ModelPostgresSystem {
        let serialized_system = builder::build_system_from_str(model_str, file_name).unwrap();
        let system = SerializableSystem::deserialize(serialized_system).unwrap();

        deserialize_postgres_subsystem(system)
    }

    fn deserialize_postgres_subsystem(system: SerializableSystem) -> ModelPostgresSystem {
        system
            .subsystems
            .into_iter()
            .find_map(|subsystem| {
                if subsystem.id == "postgres" {
                    Some(ModelPostgresSystem::deserialize(
                        subsystem.serialized_subsystem,
                    ))
                } else {
                    None
                }
            })
            // If there is no database subsystem in the serialized system, create an empty one
            .unwrap_or_else(|| Ok(ModelPostgresSystem::default()))
            .unwrap()
    }
}
