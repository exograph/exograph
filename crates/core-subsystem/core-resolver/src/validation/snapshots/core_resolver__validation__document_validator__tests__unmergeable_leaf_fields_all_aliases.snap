---
source: crates/core-subsystem/core-resolver/src/validation/document_validator.rs
expression: validator.validate(create_query_document(query))
---
Err(
    MergeFields(
        "t",
        [
            Pos(4:21),
            Pos(6:21),
        ],
    ),
)
